keps/sig-node/277-ephemeral-containers/README.md

```diff
@@ -675,60 +663,113 @@ _This section must be completed when targeting alpha to a release._
-  The e2e framework does not currently support enabling or disabling feature
-  gates. However, unit tests in each component dealing with managing data, created
-  with and without the feature, are necessary. At the very least, think about
-  conversion tests if API types are being modified.
+
+  Some unit tests are exercised with the feature both enabled and disabled to
+  verify proper behavior in both cases. Integration test verify that the API
+  server accepts/rejects requests when the feature is enabled/disabled.
+
 
 ### Rollout, Upgrade and Rollback Planning
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can a rollout fail? Can it impact already running workloads?**
-  Try to be as paranoid as possible - e.g., what if some components will restart
-   mid-rollout?
+
+  This feature allows setting a new field, `ephemeralContainers` in a Pod spec.
+  Enabling the feature won't affect existing workloads since they were not
+  previously allowed to set this field.
+
+  Component restarts won't affect this feature.
 
 * **What specific metrics should inform a rollback?**
 
+  A rollback is only indicated if there's a catastrophic failure that prevents
+  the cluster from functioning normally, for example if pod or container
+  creation begins to fail.
+
 * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
   Describe manual testing that was done and the outcomes.
   Longer term, we may want to require automated upgrade/rollback tests, but we
   are missing a bunch of machinery and tooling and can't do that now.
 
+  Since this feature is not critical to production workloads, the main risk
+  is that enabling the feature will adversely affect other components when
+  it becomes enabled by default.
+
+  This feature is implemented in the apiserver and kubelet. The change to the
+  core API has had several releases to soak. For the apiserver, the main risk
+  is described by [Adding Unstable Features to Stable Versions]. Specifically:
+
+  > Ensuring existing data is preserved is needed so that when the feature is
+  > enabled by default in a future version n and data is unconditionally allowed
+  > to be persisted in the field, an n-1 API server (with the feature still
+  > disabled by default) will not drop the data on update.
+
+  We've followed the instructions in this doc for how to persist unstable fields
+  during an update. The case for this feature is slightly more complicated
+  because the field may not be set by the default update resource. This logic
+  has been in place for several releases.
+
+  The apiserver and kubelet have existing upgrade tests
+  (https://testgrid.k8s.io/google-gce-upgrade), but these likely don't exercise
+  ephemeral containers. We'll investigate whether it's possible to add ephemeral
+  containers to these existing tests.
+
+  TODO(verb): create issue to update existing upgrade tests after discussing
```

Comment:
You say you followed the instructions, but do you have unit tests for it, and manual testing to show that after rolling back there is no problem, and after rolling forward again what happens?

Reply:
Yes, this behavior is tested in [TestDropEphemeralContainers](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/api/pod/util_test.go#L1450)

I tend to mix up upgrade/downgrade and feature enable/disable testing. enable/disable testing is the thing I care most about testing, and it matters in context of upgrades when the feature flag default changes. I realize now this section that I wrote in the "upgrade->downgrade->upgrade" path is more correctly about "disabled->enabled->disabled". I'm not sure if this was your intent for this section.

Anyway, I've done feature enablement testing manually, and I've just run an extra "disabled at 1.20 -> enabled at 1.21-alpha -> disabled at 1.20 -> enabled at 1.21-alpha" test, and everything looks ok.

I haven't done "enabled at 1.20 -> enabled at 1.21-alpha" testing since implementation hasn't changed in several releases, but I can if you'd like.

I've moved the bit about unstable fields to the enable/disable feature section and added descriptions of the unit and manual tests.